content/blog/sivayanama-testing-opsramp-apis-with-pytest-fixtures.md

```diff
@@ -0,0 +1,242 @@
+---
+title: Testing OpsRamp APIs with PyTest Fixtures
+date: 2023-07-26T06:58:31.750Z
+featuredBlog: true
+author: BalaSubramanian Vetrivel
+authorimage: /img/vetrivel-balasubramanian1-photo.jpg
+disable: false
+---
+<!--StartFragment-->
+
+# Testing OpsRamp APIs with PyTest Fixtures
+
+
+
+OpsRamp platform  provides rich set of APIs. By using these APIs customers can build soultion to automate various ITOM workflows such as discovery and monitoring , event and incident mangement and remidiation and automation. Testing these APIs, automated solution  are very critical to ensure reliability and stability of the same. PyTest is a powerful python testing framework and it is widely used to test APIs. Fixtures is one of the most important capabilities of PyTest framework. In this article, we will discuss some advanced techniques of testing OpsRamp APIs using PyTest fixtures.
+
+
+
+## What Fixtures are 
+
+PyTest fixtures are a special type of python function that provision fixed base line for testing.With the help of this base line we can ensure tests are run in reliable manner and produce consistent results and the same tests can be repeatable.
+
+## Install PyTest
+
+Run the below command 
+
+```shell
+pip install -U pytest
+```
+
+ 
+
+## Verify PyTest installation 
+
+Run the below command 
+
+```shell
+pytest --version
+pytest 7.4.0
+
+```
+
+## Define Fixture
+
+We can define fixtures just by decorating a simple python function with [@pytest.fixture](https://docs.pytest.org/en/6.2.x/reference.html#pytest.fixture) for example 
+
+```python
+import pytest
+
+@pytest.fixture
+def hello_world():
+    return 'Hello, Happy testing'
+```
+
+
+
+## Invoke fixtures
+
+Test functions can invoke fixtures just by declaring the required fixtures as arguments. 
+
+```python
+def test_hello_world(hello_world):
+    assert hello_world == 'Hello, Happy testing'
+```
+
+## Invoke OpsRamp API
+
+Let us try to execute OpsRamp [Get Access Token API](https://develop.opsramp.com/v2/api/auth/tenancy-auth-oauth-token)  as shown below.
+
+```python
+import http.client
+import json
+import traceback
+import urllib
+from asyncio.log import logger
+
+import pytest
+
+def get_auth_header(api_endpoint, client_id, client_secret, grant_type="client_credentials"):
+    """
+    Returns bearer token string in the blow format
+    bearer adfa-afdaf-1599-402c-a3ee-1ed24f597cc8
+    """
+
+    api_connection = http.client.HTTPSConnection(api_endpoint)
+    params = {"client_id": client_id, "client_secret": client_secret, "grant_type": "client_credentials"}
+
+    payload = urllib.parse.urlencode(params)
+
+    headers = {
+        'accept': "application/json",
+        'content-type': "application/x-www-form-urlencoded",
+    }
+
+```
+
+## Define Fixture to get Access Token
+
+You can transform above function as PyTest Fixture by simply decorating the code as shown below 
+
+
+
```

Comment:
Suggest to make the two coding lines below as code block snippet.
Or if you don't want to use code block snippet for these two code lines, make sure to "escape" special character (with a backslash before the underscore) for the function get_auth_header (in the preview the underscore characters are not rendered for the function get_auth_header.
See the tip #13 in the CMS editor guide: https://github.com/hpe-dev-incubator/hpe-dev-portal/blob/master/docs/ContributorGuide-v2.md#tips-and-trick-using-the-cms-editor